File: freighter/py/tests/interface.py
#  Copyright 2026 Synnax Labs, Inc.
#
#  Use of this software is governed by the Business Source License included in the file
#  licenses/BSL.txt.
#
#  As of the Change Date specified in that file, in accordance with the Business Source
#  License, use of this software will be governed by the Apache License, Version 2.0,
#  included in the file licenses/APL.txt.

from __future__ import annotations

from pydantic import BaseModel

from freighter import ExceptionPayload, register_exception


class Message(BaseModel):
    id: int | None
    message: str | None

    @classmethod
    def new(cls: type[Message]) -> Message:
        return Message(id=None, message=None)


class Error(Exception):
    code: int
    message: str

    def __init__(self, code: int, message: str):
        self.code = code
        self.message = message
        super().__init__(message)


def encode_test_error(exc: Exception) -> ExceptionPayload:
    if not isinstance(exc, Error):
        raise TypeError
    assert isinstance(exc, Error)
    return ExceptionPayload(type="integration.error", data=f"{exc.code},{exc.message}")


def decode_test_error(encoded: ExceptionPayload) -> Exception | None:
    if encoded.type != "integration.error":
        return None
    data = encoded.data
    assert data is not None
    code, message = data.split(",")
    return Error(int(code), message)


register_exception(encode_test_error, decode_test_error)

File: freighter/integration/http/interface.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package http

import (
	"context"
	"strconv"
	"strings"

	"github.com/synnaxlabs/freighter"
	"github.com/synnaxlabs/x/errors"
)

type Message struct {
	Message string `json:"message" msgpack:"message"`
	ID      int    `json:"id" msgpack:"id"`
}

type (
	Server       = freighter.StreamServer[Message, Message]
	Client       = freighter.StreamClient[Message, Message]
	ServerStream = freighter.ServerStream[Message, Message]
	ClientStream = freighter.ClientStream[Message, Message]
)

type TestError struct {
	Message string
	Code    int
}

func (t TestError) Error() string {
	return t.Message
}

func encodeTestError(_ context.Context, err error) (errors.Payload, bool) {
	var errTest TestError
	ok := errors.As(err, &errTest)
	if !ok {
		return errors.Payload{}, false
	}
	return errors.Payload{
		Type: "integration.error",
		Data: strconv.Itoa(errTest.Code) + "," + errTest.Message,
	}, true
}

func decodeTestError(_ context.Context, pld errors.Payload) (error, bool) {
	if pld.Type != "integration.error" {
		return nil, false
	}
	parts := strings.Split(pld.Data, ",")
	if len(parts) != 2 {
		return errors.New("unexpected error format"), true
	}
	code, err := strconv.Atoi(parts[0])
	if err != nil {
		return err, true
	}
	return TestError{
		Code:    code,
		Message: parts[1],
	}, true
}

func init() {
	errors.Register(encodeTestError, decodeTestError)
}
